### PR TITLE
enh(docs) Make demo more static

### PR DESCRIPTION
**Pull request by Hirse**
This PR updates the [`demo.html`](https://highlightjs.org/static/demo/) page by moving as much of the initialization as possible to the build step.

### Changes
- Remove dependency on jQuery
  - Move jQuery and Vue files to `tools/vendor` as they are still required for `developer.html`
- Static generation
  - Generate list of categories during build
  - Generate list of themes during build
  - Set language categories (`common`, `all`) during build
- Use css-`flex` instead of JS to set the height of the side bar lists
- Use WebWorker for highlighting where possible
- Use favicon from website
- Lazily load theme files on selection by setting `disabled` attribute on `link`

### Checklist
- [x] ~Added markup tests~ NA
- [x] ~Updated the changelog at `CHANGES.md`~
- [x] Added myself to `AUTHORS.txt`, under Contributors

### CODE REVIEW

File: demo/demo.js
Comment (joshgoebel) [inline demo/demo.js:58]: Would `alternative` be better than disabled?  I don't really know but it's what I've used for this kind of thing in the past.

Comment (joshgoebel) [reply demo/demo.js:58]: Don't we still have to toggle `alternative` on/off?

**Comment by joshgoebel** [comment]
So I think I agree with the overall goals/ambition here (more static rendering, throw out jQuery, etc)... BUT:

_Why all the Worker complexity and not just render super-simpler HTML and call our default init function to do all the heaving lifting?_ 

What do we gain with all that custom JS you've written vs almost no JS and just leaning on the libraries existing "just do it" auto support?

File: demo/demo.js
Comment (Hirse) [reply demo/demo.js:58]: I can have a look, but here I just followed what the previous script did:
https://github.com/highlightjs/highlight.js/blob/7e108baaa3be0a7ad402d60259d060fcefa44021/demo/demo.js#L86-L90

**Comment by Hirse** [comment]
I added the worker mostly because I saw it on the [`/usage`](https://highlightjs.org/usage) page, thought it looked cool, and wanted to try it out.

Switched to a simple `initHighlightingOnLoad` for now.

>     * Why not just considering rendering everything server-side... why run Highlight.js on the client at all for the demo?

I had that thought too, but was wondering whether the demo should also demo the client-side usage or only the results?

File: demo/demo.js
Comment (Hirse) [reply demo/demo.js:34]: The current usage (or `alternative`/`disabled`) works fine in Firefox and Chromium. If you are on Mac, could you try it for Safari?

Comment (joshgoebel) [reply demo/demo.js:34]: Ah, I was looking more closely at the new code than the old... 

From MDN (learning something new):

> Alternate (rel="alternate stylesheet", title="..." must be specified): disabled by default, can be selected.

So alternates are simply disabled by default... I've always toggled their "alternate" status (making them active by default), but seems you can just as easily toggle their "disabled" status, thereby also making them active. :-)

So this seems a fine enough strategy.

File: demo/style.css
Comment (joshgoebel) [inline demo/style.css:70]: Nice!

**Comment by joshgoebel** [comment]
> I added the worker mostly because I saw it on the /usage page, thought it looked cool, and wanted to try it out.

Trying new things is fun.  What did you think?  I think it's a bit annoying and wonder if long-term we shouldn't just make it much easier to use workers "out of the box".  No one is really raising any issues though so I haven't considered it a high priority... and seems less relevant than ever since all the recent catastrophic regex fixes... or maybe one benefit of workers is that the implementor writes all the code and the library is *entirely* isolated inside the worker process.

Any thoughts on big picture Worker stuff?

> Switched to a simple initHighlightingOnLoad for now.

👍🏼  You should really consider joining our Slack.

> wondering whether the demo should also demo the client-side usage or only the results?

*shrugs*  If that is the argument though then I think it's also an argument for using the simpler `initHighlightingOnLoad`, so I think it's good that we simplified that.  Personally I think the demo is intended to show *visually* what our library does, not be an example of usage (we have plenty of that in the README).

No need to hold up this PR - and I don't think we HAVE to server-side render by any means... but I don't think there is any strong reason that the demo highlighting portion MUST be rendered on the client.

File: demo/demo.js
Comment (joshgoebel) [inline demo/demo.js:1]: Why debug mode?

File: demo/index.html
Comment (joshgoebel) [inline demo/index.html:5]: Why? HTML5 does not need to be valid XHTML nor is there a good reason to force it to be off the top of my head.  Is there a big reason for these changes?

File: demo/demo.js
Comment (Hirse) [reply demo/demo.js:1]: Because it was before, but no preference either way.
https://github.com/highlightjs/highlight.js/blob/4d9d9e1832e3a8ce48fc116e2aee1a5225351bdc/demo/demo.js#L4

File: demo/index.html
Comment (Hirse) [reply demo/index.html:5]: No, mostly personal preference.

**Comment by Hirse** [comment]
> 👍🏼 You should really consider joining our Slack.

The button in the Readme doesn't seem to work anymore.
[slack](https://join.slack.com/t/highlightjs/shared_invite/zt-jatdlkw4-h3LdjU5rC23t7aQ6zqoxzw)

File: demo/demo.js
Comment (joshgoebel) [reply demo/demo.js:1]: Oh, hmmm... seems like it shouldn't really be needed... but if the page every blew up it would be easier to diagnose...  I guess if it was there before we can just defer to that then.

File: demo/index.html
Comment (joshgoebel) [reply demo/index.html:5]: I like this [answer from StackOverflow](https://stackoverflow.com/questions/3558119/are-non-void-self-closing-tags-valid-in-html5):

> **In HTML5, the meaning of <foo /> depends on the type of element.**

> On HTML elements that are designated as void elements (essentially "An element that existed before HTML5 and which was forbidden to have any content"), end tags are simply forbidden. The slash at the end of the start tag is allowed, but has no meaning. It is just syntactic sugar for people (and syntax highlighters) that are addicted to XML.

Since we're using HTML5 not XHTML I think we should be clear and just write clean and simple HTML5, not XHTML.  If we really wanted to write XHTML we should change our DTD to say so, but I don't see any benefit to this.

**Comment by joshgoebel** [comment]
Try again, updated the link.  It expires every 30 days. :(

File: demo/index.html
Comment (Hirse) [reply demo/index.html]: Removed.

File: demo/index.html
Comment (joshgoebel) [inline demo/index.html:7]: Long-term should we considering pulling this into our repo itself vs linking to the website? Not sure i have a strong opinion.

File: tools/build_browser.js
Comment (joshgoebel) [inline tools/build_browser.js:78]: Can all this be pulled out into a `renderIndex()` function?  installDemo is getting a bit long.

**Review by joshgoebel — CHANGES_REQUESTED** [review]
Looks good other than my one comment about factoring out `renderIndex()`... I'll circle back later actually test it locally before merging.

File: demo/index.html
Comment (Hirse) [reply demo/index.html:7]: I don't think it's necessary to pull it into the repo as it's only relevant once the demo is part of the website. We could use a relative path (`../favicon.png`), but that's not really better (and doesn't show up locally).

Comment (joshgoebel) [reply demo/index.html:7]: Then we'll leave it as is for now.

File: tools/build_browser.js
Comment (Hirse) [reply tools/build_browser.js:82]: Done.

**Comment by joshgoebel** [comment]
In the future perhaps please just add new commits as you go along... GitHub makes it much easier to review the changes (ie, just the new commits)... when you rewrite the whole history the review screen forces me to start over and look at *everything single change*.

We squash almost all commit history anyways in PRs so there is no need for the history to be "clean".

**Comment by joshgoebel** [comment]
Screen Shot 2020-12-18 at 1 09 13 AM

misc (undefined)

Did we break something?

**Comment by Hirse** [comment]
common, misc, and all are hard-coded because they have a specific place in the order. Added another filtering to only show them if there are languages for the category:
- `node ./tools/build.js vbnet`
  image
- `node ./tools/build.js :common`
image

**Comment by joshgoebel** [comment]
Yea. Clean and simple fix.

File: demo/style.css
Comment (joshgoebel) [inline demo/style.css:39]: Where are we using tabs that these needed to be added?

File: demo/style.css
Comment (joshgoebel) [inline demo/style.css:36]: Isn't this already coming from the themes?

File: demo/style.css
Comment (joshgoebel) [inline demo/style.css]: I'm not sure if we're fighting over styling here or if this is accidental.  I _very specifically_ tweaked this last time to add more padding... and almost didn't notice that you removed this... I've added it back.

Comment (Hirse) [reply demo/style.css]: The top padding is compatible with school-book, but the left padding is not:
image

Having school-book style padding on other themes feels a bit odd (especially since it's only on the top and left)
image
image

I would suggest letting the theme set the padding.

File: demo/style.css
Comment (Hirse) [reply demo/style.css:36]: Ah, yes. I added it here while I was highlighting asynchronously with the worker so the code blocks don't jump around.

File: demo/style.css
Comment (Hirse) [reply demo/style.css:39]: 1c, autoit, crmsh, dts, gml, hsp, makefile, roboconf, sqf, tp

For example:
https://github.com/highlightjs/highlight.js/blob/3a89dfb51006751b34501ea4f78d744b6d98e91a/test/detect/1c/default.txt#L3

File: demo/style.css
Comment (joshgoebel) [reply demo/style.css]: > Having school-book style padding on other themes feels a bit odd  

I don't find it problematic.  As a general rules themes should not enforce pixel-precise padding (and no other themes do).  I may even remove school book in v11 (or remove it's BG at least).  If a modern theme were submitted that did something similar I would reject it.

> I would suggest letting the theme set the padding.

No, sorry, the default is pretty awful IMHO.  I think I'll add it to the list of things to change in v11.

File: demo/style.css
Comment (joshgoebel) [reply demo/style.css:39]: Ok, I was not aware of this.  And how did we decide on 4 as the proper tab width?  What is the default?

File: demo/style.css
Comment (joshgoebel) [reply demo/style.css:36]: But the CSS should have nothing to do with the worker...?

**Comment by joshgoebel** [comment]
If there is some way to override all the themes padding (other than schoolbook) without custom JS (which is overkill) I'd be open to that... ie, some sort of CSS that I'm not aware of off the top of my head... but I don't think that's easily doable?

File: demo/style.css
Comment (joshgoebel) [reply demo/style.css:39]: I wonder if those languages prefer/use tabs over spaces or if this is just a legacy oddity... This is a spaces codebase and my first urge was to go and change them all to spaces. :)

Comment (Hirse) [reply demo/style.css:39]: I just picked 4 as it tends to look better than 8 and seems to usually be the default in most editors.

**Comment by Hirse** [comment]
> If there is some way to override all the themes padding (other than schoolbook) without custom JS (which is overkill) I'd be open to that... ie, some sort of CSS that I'm not aware of off the top of my head... but I don't think that's easily doable?

We could set the name of the current theme as class on the `body` (or `main`) which would allow more specific selectors, but that does need a tiny bit of JS and I don't think it's really worth it.

**Review by joshgoebel — APPROVED** [review]

**Pull request merged**
